src/qtcore/qml/lib/parser.js

```diff
@@ -293,20 +293,104 @@ function js_error(message, line, col, pos, comment) {
 };
-
-function extractLinesForErrorDiag(text, line)
-{
-  var r = "";
-  var lines = text.split("\n");
-
-  for (var i = line - 3; i <= line + 3; i++)
-  if (i >= 0 && i < lines.length ) {
-      var mark = ( i == line ) ? ">>" : "  ";
-      r += mark + i + "  " + lines[i] + "\n";
-  }
-
-  return r;
-}
+// CJK characters in the console display longer lengths, calculate the length of a string of actual display
+var normal_c = " ";
+var cjk_c = "　";
+var cjk = function(char_code) {
+    return 0x4E00 <= char_code && char_code <= 0x9FFF ||
+        0x3400 <= char_code && char_code <= 0x4DFF ||
+        0x20000 <= char_code && char_code <= 0x2A6DF ||
+        0xF900 <= char_code && char_code <= 0xFAFF ||
+        0x2F800 <= char_code && char_code <= 0x2FA1F;
+};
+function replacerConsoleRange(str) {
+    var res = "";
+    for (var i = 0, len = str.length, cc; i < len; i += 1) {
+        cc = str.charCodeAt(i);
+        if (cc === 9) { // \t
+            res += '\t' // tab不管，否则不同的控制台下tab的长度是不一样的；除非在源码那边也进行修改
+        } else if (cjk(cc)) {
+            res += cjk_c
+        } else {
+            res += normal_c
+        }
+    }
+    return res;
+};
+function repeatString(_str, count) {
+    'use strict';
+    if (_str == null) {
+        throw new TypeError('can\'t convert ' + _str + ' to object');
+    }
+    var str = '' + _str;
+    count = +count;
+    if (count != count) {
+        count = 0;
+    }
+    if (count < 0) {
+        throw new RangeError('repeat count must be non-negative');
+    }
+    if (count == Infinity) {
+        throw new RangeError('repeat count must be less than infinity');
+    }
+    count = Math.floor(count);
+    if (str.length == 0 || count == 0) {
+        return '';
+    }
+    // Ensuring count is a 31-bit integer allows us to heavily optimize the
+    // main part. But anyway, most current (August 2014) browsers can't handle
+    // strings 1 << 28 chars or longer, so:
+    if (str.length * count >= 1 << 28) {
+        throw new RangeError('repeat count must not overflow maximum string size');
+    }
+    var rpt = '';
+    for (;;) {
+        if ((count & 1) == 1) {
+            rpt += str;
+        }
+        count >>>= 1;
+        if (count == 0) {
+            break;
+        }
+        str += str;
+    }
+    // Could we try:
+    // return Array(count + 1).join(_str);
+    return rpt;
+};
+function extractLinesForErrorDiag(code_text, line, column, options) {
+    options || (options = {});
+    var line_range = options.line_range << 0 || 3;
+
+    var codeLines = code_text.split("\n");
+
+    // Error line number index
+    var show_line_index = line;
+    // The starting line of the display code number
+    var show_start_line = Math.max(show_line_index - line_range, 0);
+    // End line of code displays the number of
+    var show_end_line = show_line_index + line_range;
+
+    // The number of characters required to display line, blank spaces + 1
```

Comment:
it took me some time to understand, that you mean "line number" here. Maybe you can add it for clarity. :)